File: src/webhook.ts
/**
 * Incoming Webhook Event from Pinto Bot Platform
 */
export interface WebhookBaseEvent {
  event: 'ping' | 'message.created' | string
  timestamp: string
}

export interface WebhookPingEvent extends WebhookBaseEvent {
  event: 'ping'
  message: string
}

export interface WebhookMessageEvent extends WebhookBaseEvent {
  event: 'message.created'
  bot_id: string
  chat_id: string
  message_id: string
  sender: {
    user_id: string
    name: string
  }
  message: string
}

export type WebhookEvent = WebhookPingEvent | WebhookMessageEvent | (WebhookBaseEvent & Record<string, unknown>)

export interface WebhookReplyPayload {
  reply_message: string
  media_url?: string
}

/**
 * Verify Webhook Secret using timing-safe comparison
 * @param headerSecret The X-Pinto-Secret header value from incoming request
 * @param configuredSecret The Webhook Secret configured in Developer Portal
 */
export function verifyWebhookSecret(headerSecret: string | null | undefined, configuredSecret: string): boolean {
  if (!headerSecret || !configuredSecret) return false
  if (headerSecret.length !== configuredSecret.length) return false

  let mismatch = 0
  for (let i = 0; i < headerSecret.length; i++) {
    mismatch |= headerSecret.charCodeAt(i) ^ configuredSecret.charCodeAt(i)
  }
  return mismatch === 0
}

/**
 * Parse incoming Webhook payload
 */
export function parseWebhookEvent(rawBody: string | Record<string, unknown>): WebhookEvent {
  if (typeof rawBody === 'string') {
    return JSON.parse(rawBody) as WebhookEvent
  }
  return rawBody as WebhookEvent
}

/**
 * Helper to construct a Webhook Reply response payload
 */
export function createReplyResponse(replyMessage: string, mediaUrl?: string): WebhookReplyPayload {
  const res: WebhookReplyPayload = { reply_message: replyMessage }
  if (mediaUrl) {
    res.media_url = mediaUrl
  }
  return res
}

File: go/webhook.go
package pinto

import (
	"crypto/subtle"
	"encoding/json"
	"errors"
	"io"
	"net/http"
)

// VerifyWebhookSecret validates incoming X-Pinto-Secret header against configured secret
// using constant-time comparison to prevent timing attacks.
func VerifyWebhookSecret(headerSecret, configuredSecret string) bool {
	if headerSecret == "" || configuredSecret == "" {
		return false
	}
	return subtle.ConstantTimeCompare([]byte(headerSecret), []byte(configuredSecret)) == 1
}

// ParseWebhookEvent parses an incoming HTTP request from Pinto Bot Webhook
func ParseWebhookEvent(r *http.Request) (*WebhookEvent, error) {
	if r == nil || r.Body == nil {
		return nil, errors.New("pinto: request or body is nil")
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		return nil, err
	}

	var event WebhookEvent
	if err := json.Unmarshal(body, &event); err != nil {
		return nil, err
	}

	return &event, nil
}

// NewReplyResponse creates a Webhook reply payload
func NewReplyResponse(message string, mediaURL ...string) *WebhookReply {
	reply := &WebhookReply{
		ReplyMessage: message,
	}
	if len(mediaURL) > 0 && mediaURL[0] != "" {
		reply.MediaURL = mediaURL[0]
	}
	return reply
}
